src/melee.cpp

```diff
@@ -422,17 +424,39 @@ void Character::melee_attack( Creature &t, bool allow_special, const matec_id &f
             return;
         }
     }
-    item &cur_weapon = allow_unarmed ? used_weapon() : weapon;
+    item *cur_weapon = allow_unarmed ? &used_weapon() : &weapon;
+
+    // If no weapon is selected, use highest layer of gloves instead.
+    bool unarmed_flag_set = false;
+    if( cur_weapon->is_null() ) {
+        bool found_glove = false;
+        for( item &worn_item : worn ) {
+            // Uses enum layer_level to make distinction for top layer.
+            if( ( worn_item.covers( bodypart_id( bp_hand_l ) ) &&
+                  worn_item.covers( bodypart_id( bp_hand_r ) ) ) ) {
+                if( cur_weapon->is_null() || ( worn_item.get_layer() >= cur_weapon->get_layer() ) ) {
+                    cur_weapon = &worn_item;
+                    found_glove = true;
+                }
+            }
+        }
+
+        // Treat all gloves as unarmed weapons.
+        if( found_glove && !cur_weapon->has_flag( flag_UNARMED_WEAPON ) ) {
```

Comment:
You can remove the "has_flag" check as redundant. If an item already has this flag, "set_flag" below will simply do nothing.

Reply:
> You can remove the "has_flag" check as redundant. If an item already has this flag, "set_flag" below will simply do nothing.

It's `not has_flag`.

Reply:
The logic is: if item doesn't have the flag, add it. The check could be omitted - we're adding flag anyways, and if there's already flag exist, adding flag just won't do anything.

Reply:
Right. Now I understand what you wrote before.